README
README for RudeDatabase
See the end of file for copying conditions

This file contains the following sections, please
glance through them all before installing RudeDatabase.

	* Introduction
	* Bug Reporting
	* Copying
	

*** INTRODUCTION ***

RudeDatabase is a library that provides a simple interface
to the mysql database library.

Documentation, discussions, and new releases are available at 

http://www.rudeserver.com/

See file `COPYING' for copying conditions.
See file `INSTALL' for compilation and installation instructions.
See file `NEWS' for a list of major changes in the current release.
See file `THANKS' for a list of contributors.
See file 'AUTHORS' for contact information


*** BUG REPORTING ***

Please Send bug reports to <rudedatabase@rudeserver.com>.
A bug report should contain an adequate description of the 
problem, your input, what you expected, what you got, 
and why this is wrong.  Where applicable, please submit actual data
files with your report. 

Your feedback will help us to make a better and more portable package.
Consider documentation errors as bugs, and report them as such.  If you
develop anything pertaining to RudeDatabase or have suggestions, let us know
and share your findings by writing to <rudedatabase@rudeserver.com> or
by visiting the discussion forum accessable from the RudeServer 
website at http://www.rudeserver.com/



*** COPYING ***

Copyright (C) 2002, 2003, 2004, 2005 Matthew Flood
See file AUTHORS for contact information

This file is part of RudeDatabase.

RudeDatabase is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

RudeDatabase is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with RudeDatabase; see the file COPYING.  If not, write to
the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.



NEWS
RudeDatabase NEWS - User visible changes.
See ChangeLog for more detailed information on internal changes
See the end for copying conditions.

* Tue Jan 08 2008 Matt Flood <matt@rudeserver.com>
- 4.1.0-1
- Bug fixes

____________________________________________________________________________
Release 4.0.1-1 - Matthew Flood 2007-05-29

	- Updated man page
	- minor build script modifications
	- Updated the man page
	- Changed internal namespace from database to db

____________________________________________________________________________
Release 3.0.1-1 - Matthew Flood 2007-05-29

	- created rudedatabase man page
	- minor build script modifications

____________________________________________________________________________
Release 3.0.0-1 - Matthew Flood 2005-06-10

	** RudeDatabase became open source and is distributed under the GNU Public License.
		Commercial licenses are available for a fee, and allow the holder
		to use the library in commercial applications. Contact info@rudeserver.com
		for more details.

	** Changed package versioning: 
		now using the LIBRARY_VERSION-RELEASE_VERSION
		Release version will be set to 1 for each new lib version
		
	** Rolled the package into binary and source RPMs
__________________________________________________________________________________


Copyright (C) 2002, 2003, 2004, 2005 Matthew Flood
See file AUTHORS for contact information

This file is part of RudeDatabase.

RudeDatabase is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

RudeDatabase is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with RudeDatabase; see the file COPYING.  If not, write to
the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.



src/database.cpp
// database.cpp: interface for the database class.
//
// Copyright (C) 2002, 2003, 2004, 2005 Matthew Flood
// See file AUTHORS for contact information
//
// This file is part of RudeDatabase.
//
// RudeDatabase is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2, or (at your option)
// any later version.
// 
// RudeDatabase is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with RudeDatabase; (see COPYING) if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
// 02111-1307, USA.
//------------------------------------------------------------------------


#include "database.h"

#ifndef INCLUDED_IMPLEMENTATION_H
#include "implementation.h"
#endif

#ifndef INCLUDED_STRING
#include <string>
#define INCLUDED_STRING
#endif

namespace rude{

std::string invalid_context = "Invalid Context";

std::map<std::string, Database*> Database::s_databases;


//////////////////////////////////////////////////////////
// Static Method definitions
//////////////////////////////////////////////////////////
Database *Database::instance(const char* contextname)
{
	std::string context=contextname? contextname: "";
	Database *database = s_databases[context];
	
	if(!database)
	{
		invalid_context = "Invalid Context '";
		invalid_context += context;
		invalid_context += "'";
		
		// The context they asked for does not exist!!!
		// pretend it does exist, (create a new database)
		// but give it an empty (NULL) implementation
		//
		database = new Database(0);

		// add it to map so other calls to same context are handled
		// without having to re-create object.
		//
		s_databases[context]=database;
	}
	
	// Database should connect here if it is not already connected!!!!
	//
	database->connect();

	return database;
}

void Database::finish()
{
	std::map<std::string, Database*>::iterator iter;
	for(iter = s_databases.begin(); iter != s_databases.end(); iter ++)
	{
		Database *d = (*iter).second;
		if(d)
		{
			delete d;
		}
	}
	s_databases.clear();
}

void Database::addContext(const char *contextname, const char *databasename, 
	const char *username, const char *password, const char *server, int port)
{
	// Database should not connect here!!!!
	//
	
	rude::db::Implementation *implementation = new rude::db::Implementation(  username, 
																			password, 
																			server, 
																			port, 
																			databasename);
	Database *newdatabase = new Database(implementation);
	std::string context=contextname ? contextname : "";

	// if we are replacing existing contexts, we
	// need to delete the old database objects first.
	//
	Database *olddatabase = s_databases[context];
	if(olddatabase)
	{
		delete olddatabase;
	}
	s_databases[context]=newdatabase;
}

char *Database::escape(const char *oldstring, int length)
{
	return rude::db::Implementation::escape(oldstring, length);
}

/////////////////////////////////////////////////////////
// Constructors & Destructors
/////////////////////////////////////////////////////////

Database::Database(rude::db::Implementation *implementation)
{
	d_implementation = implementation;
}

Database::~Database()
{	
	// If database is connected, 
	// deleting implementation should disconnect....
	//
	if(d_implementation)
	{
		delete d_implementation;
		d_implementation=0;
	}
}

//////////////////////////////////////////////////////////
// Instance Method Definitions
//////////////////////////////////////////////////////////

const char *Database::getError() const
{
	return d_implementation ? d_implementation->getError() : invalid_context.c_str();
}

const char *Database::stat()
{
	return d_implementation ? d_implementation->stat() : invalid_context.c_str();
}


RUDEBOOL Database::storeResultQuery(const char *query)
{
	return d_implementation ? d_implementation->storeResultQuery(query) : false;
}
	
RUDEBOOL Database::useResultQuery(const char *query)
{	
	return d_implementation ? d_implementation->useResultQuery(query) : false;
}

RUDEBOOL Database::nextRow()
{
	return d_implementation ? d_implementation->nextRow() : false;
}

RUDEBOOL Database::connect()
{
	return d_implementation ? d_implementation->connect() : false;
}

const char *Database::column(int index)
{
	return d_implementation ? d_implementation->column(index) : 0;
}

RUDEBOOL Database::freeResult()
{
	return d_implementation ?  d_implementation->freeResult() : false;
}

const char *Database::singleValueQuery(const char *query)
{
	return d_implementation ? d_implementation->singleValueQuery(query) : 0;
}

RUDEBOOL Database::executeQuery(const char *query)
{
	return d_implementation ? d_implementation->executeQuery(query) : false;
}

int Database::insertQuery(const char *query)
{
	return d_implementation ? d_implementation->insertQuery(query) : 0;
}


} // end namespace



src/implementation.h
// implementation.h
//
// Copyright (C) 2002, 2003, 2004, 2005 Matthew Flood
// See file AUTHORS for contact information
//
// This file is part of RudeDatabase.
//
// RudeDatabase is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2, or (at your option)
// any later version.
// 
// RudeDatabase is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with RudeDatabase; (see COPYING) if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
// 02111-1307, USA.
//------------------------------------------------------------------------


#ifndef INCLUDED_RUDE_DATABASE_IMPLEMENTATION_H
#define INCLUDED_RUDE_DATABASE_IMPLEMENTATION_H


#ifndef INCLUDED_STRING
#include <string>
#define INCLUDED_STRING
#endif


#ifndef WIN32
#ifndef INCLUDED_MYSQL_H
#include <mysql/mysql.h>
#define INCLUDED_MYSQL_H
#endif
#endif

// These header files need to be included last,
// for some odd reason...
//
#ifdef WIN32
#include "my_global.h"
#include "mysql.h"#endif

#ifndef RUDE_DB_BOOL
#ifdef WIN32
typedef unsigned int RUDEBOOL;
#else
typedef bool RUDEBOOL;
#endif
#define RUDE_DB_BOOL
#endif


namespace rude{
namespace db{

class Implementation 
{
	std::string d_username;
	std::string d_password;
	std::string d_server;
	std::string d_database;
	int d_port;

	std::string d_error;

	RUDEBOOL d_connected;
	
	MYSQL *d_conn;
	MYSQL_RES *d_result;
	MYSQL_ROW d_row;
	std::string d_simplereturn; // stores simple return data...
	
protected:
	
	void setError(const char * error);
public:

	static char *escape(const char *string, int length);

	
	Implementation(const char *username, 
						const char *password, 
						const char *server, 
						int port, 
						const char *database);
	
	~Implementation();
	
	RUDEBOOL connect();
	RUDEBOOL disconnect();
	
	const char *stat();
	const char *getError() const;

	RUDEBOOL executeQuery(const char *query);
	const char *singleValueQuery(const char *query);
	int insertQuery(const char *query);

	RUDEBOOL storeResultQuery(const char *query);
	RUDEBOOL useResultQuery(const char *query);

	RUDEBOOL nextRow();
	const char *column(int index);
	RUDEBOOL freeResult();
};

}} // end namespaces

#endif



src/implementation.cpp
// implementation.cpp
//
// Copyright (C) 2002, 2003, 2004, 2005 Matthew Flood
// See file AUTHORS for contact information
//
// This file is part of RudeDatabase.
//
// RudeDatabase is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2, or (at your option)
// any later version.
// 
// RudeDatabase is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with RudeDatabase; (see COPYING) if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
// 02111-1307, USA.
//------------------------------------------------------------------------



#include "implementation.h"


using namespace std;
using namespace rude::db;

namespace rude{
namespace db{

//////////////////////////////////////////
// Static Variable Definitions
//////////////////////////////////////////

//////////////////////////////////////////
// Static Method Definitions
//////////////////////////////////////////


char *Implementation::escape(const char *string, int length)
{
	char *retval = new char[length * 2 + 1];
	mysql_escape_string(retval,string,length);
	return retval;
}


//////////////////////////////////////////
// Constructors & Destructor
//////////////////////////////////////////


Implementation::Implementation(const char *username, const char *password, const char *server, int port, const char *database)
{
	d_error="";
	d_username = username ? username : "";
	d_password = password ? password : "";
	d_server = server ? server : "";
	d_port = port ? port : 3306;
	d_database = database ? database : "";

	d_connected = false;
	
	d_conn=0;
	d_result=0;
	d_row=0;
	
	d_simplereturn="";
}


Implementation::~Implementation()
{
	freeResult();
	disconnect();
}


//////////////////////////////////////////
// Instance Method Definitions
//////////////////////////////////////////


void Implementation::setError(const char * error)
{
	d_error = error ? error : "";
}


const char *Implementation::getError() const
{
	return d_error.c_str();
}


RUDEBOOL Implementation::connect()
{
	if(!d_connected)
	{
		MYSQL *initconn = mysql_init(0);
		if(initconn)
		{
			MYSQL * conn = mysql_real_connect(initconn, 
						d_server.c_str() , 
						d_username.c_str(), 
						d_password.c_str(), 
						d_database.c_str(), 
						d_port, 
						0,
						0);
			if(conn)
			{
				d_conn=conn;
				d_connected=true;
			}
			else
			{
				d_conn = 0;
				setError(mysql_error(initconn));
				mysql_close(initconn);
			}
		}
		else
		{
			setError("Conn was NULL");
		}
	}
	return d_connected;
}	


RUDEBOOL Implementation::disconnect()
{
	if(d_connected)
	{
		mysql_close(d_conn);
		d_connected = false;
		return true;
	}
	else
	{
		setError("Already Disconnected");
		return false;
	}
}


const char *Implementation::stat()
{
	if(d_connected)
	{
		return mysql_stat(d_conn);
	}
	else
	{
		return "Not Connected";
	}
}


RUDEBOOL Implementation::executeQuery(const char *query)
{
	if(query)
	{
		if(d_connected)
		{
			if( mysql_query(d_conn, query) == 0)
			{
				return true;
			}
			else
			{
				setError(mysql_error(d_conn));
			}
		}
		else
		{
			setError("Not Connected");
		}
	}
	else
	{
		setError("Query to be executed was NULL");
	}
	return false;
}


int Implementation::insertQuery(const char *query)
{
	if(query)
	{
		if(d_connected)
		{
			if( mysql_query(d_conn, query) == 0)
			{
				return mysql_insert_id(d_conn);
			}
			else
			{
				setError(mysql_error(d_conn));
			}
		}
		else
		{
			setError("Not Connected");
		}
	}
	else
	{
		setError("Query to be executed was NULL");
	}
	return 0;
}

const char *Implementation::singleValueQuery(const char *query)
{	

	if(query)
	{
		if(d_connected)
		{
			if( mysql_query(d_conn, query) == 0)
			{
				//get result set
				//
				MYSQL_RES *result = mysql_store_result(d_conn);
				if(result)
				{
					//get first row
					//
					MYSQL_ROW row;
					row = mysql_fetch_row(result);
					if(row)
					{
						if(row[0])
						{
							d_simplereturn = row[0];
						}
						else
						{
							// Got a NULL
							//
							d_simplereturn = "";
						}
						mysql_free_result(result);
						return d_simplereturn.c_str();
					}
					else
					{
						setError("Could not fetch row for single value query");
					} // end if row
					mysql_free_result(result);
				}
				else
				{
					setError(mysql_error(d_conn));
				} // end if result
			}
			else
			{
				setError(mysql_error(d_conn));
			}
		} 
		else
		{
			setError("Not Connected");
		}
	}
	else
	{
		setError("Query was NULL");
	}
	return 0;
}


RUDEBOOL Implementation::freeResult()
{
	if(d_result)
	{
		mysql_free_result(d_result);
		d_result = 0;
	}
	return true;
}

RUDEBOOL Implementation::nextRow()
{
	if(d_result)
	{
		d_row = mysql_fetch_row(d_result);
		if(d_row)
		{
			return true;
		}
		setError("No More Rows");
		return false;
	}
	else
	{
		// error should have been set already.....
		return false;
	}
}

const char *Implementation::column(int index)
{
	if(d_result !=0)
	{
		if(d_row != 0)
		{
			return d_row[ index ];
		}
		else
		{
			setError("No row to work with");
		}
	}
	else
	{
		setError("No Result to Work With");
	}
	return 0;

}


RUDEBOOL Implementation::storeResultQuery(const char *query)
{
	if(d_connected)
	{
		if(query)
		{
			if( mysql_query(d_conn, query) == 0)
			{
				//get result set
				d_result = mysql_store_result(d_conn);
				if(d_result)
				{
					return true;
				}
				else
				{
					setError(mysql_error(d_conn));
				}
			}
			else
			{
				setError(mysql_error(d_conn));
			}
		}
		else
		{
			setError("Query was NULL");
		}
	}
	return false;
}


RUDEBOOL Implementation::useResultQuery(const char *query)
{
	if(d_connected)
	{
		if(query)
		{
			if( mysql_query(d_conn, query) == 0)
			{
				d_result = mysql_use_result(d_conn);
				if(d_result)
				{
					return true;
				}
				else
				{
					setError(mysql_error(d_conn));
				}
			}
			else
			{
				setError(mysql_error(d_conn));
			}
		}
		else
		{
			setError("Query was NULL");
		}
	}
	// else error already set on connect();
	return false;
}

}} // end namespaces


src/database.h
// database.h: interface for the database class.
//
// Copyright (C) 2002, 2003, 2004, 2005 Matthew Flood
// See file AUTHORS for contact information
//
// This file is part of RudeDatabase.
//
// RudeDatabase is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2, or (at your option)
// any later version.
// 
// RudeDatabase is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with RudeDatabase; (see COPYING) if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
// 02111-1307, USA.
//------------------------------------------------------------------------


#ifndef INCLUDED_DATABASE_H
#define INCLUDED_DATABASE_H


// Borland can choke on BOOLs
// so we convert to unsigned int on windows
//
#ifndef RUDE_DB_BOOL
#ifdef WIN32
typedef unsigned int RUDEBOOL;
#else
typedef bool RUDEBOOL;
#endif
#define RUDE_DB_BOOL
#endif

#ifndef	INCLUDED_MAP
#include <map>
#define INCLUDED_MAP
#endif

#ifndef INCLUDED_STRING
#include <string>
#define INCLUDED_STRING
#endif

namespace rude{

namespace db{
class Implementation;
}

class Database
{
	static std::map<std::string, Database*> s_databases;

	rude::db::Implementation *d_implementation; // Bridge
	
protected:
	//=
	// Database objects cannot be instantiated directly
	// Instead, call instance() to obtain a reference to the object. This component uses 
	// the Singleton design pattern to ensure global accessability and improve performance.
	// This architecture allows multiple parts of your program to access a particular database 
	// without having to worry about connecting or disconnecting to/from the database.  
	// If every section of your program needed to use a database and was responsible for 
	// connecting/disconnecting, then multiple connections would be opened and closed - consuming 
	// both resources and time.  Our architecture ensures that a connection is established 
	// once - and then only if necessary - and can be shared throughout the program.  Multiple 
	// contexts allow multiple connections to different databases or database accounts simultaneously.
	//=
	Database(db::Implementation *implementation);
	RUDEBOOL connect();
		
public:
	//=
	// Adds a database, username, password, server and port identified by <i>contextname</i>. 
	// The context name can be anything and is simply used as an identifier to the context.
	// After a context is added, the database represented by the context name can be obtained by
	// calling the Database::instance(contextname) method.
	// Any number of contexts can be added.  Each context should have a unique contextname.
	//= 
	static void addContext(	const char *contextname, 
									const char *database, 
									const char *username, 
									const char *password, 
									const char *server, 
									int port);
	
	//=
	// Escapes illegal characters within a string
	// This ensures that string or binary data won't mess up your queries.
	// You should escape all string data that can possibly be tainted 
	// (eg. you are not sure what is in it).
	// The caller (you) is responsible for deleting the resulting string.
	// 
	// <B>Note:</b> To avoid memory leaks, you <i>MUST</i> delete the resulting string. 
	// As such, the following sequence should always be followed when using this method:
	//
	// <code>
	// // 1. Obtain safe strings 
	// // ( as a static method, it can be called through an object or directly from the class )
	// //
	// const char *safe_string = database->escape( unsafe_string , strlen(unsafe_string) );
	// const char *safe_string2 = rude::Database::escape( another_string, strlen(another_string) );
	// 
	// // 2. ... use the safe strings
	// 
	// // 3. Delete the safe strings
	// //
	// delete [] safe_string
	// delete [] safe_string2
	// </code>
	//
	// <b>Example:</b>
	//
	//<code>
	// // Obtain database object
	// //
	// rude::Database *database = rude::Database::instance("stock");
	// 
	// const char *catname="A very ''' Awkward ===== Category Name !!'' \"\"\" ****** Indeed!!";
	// 
	// // Escape illegal characters in the new category name
	//	//
	// char *safe_name = database->escape( catname, strlen(catname));
	// 
	// // Build the SQL statement
	// // string / character data (whether escaped or not) should be surrounded by single quotes.
	// //
	//	std::string sqlstatement = "INSERT INTO tbl_category SET name='";
	//	sqlstatement += safe_name;
	//	sqlstatement += "'";
	// 
	// // REMEMBER to delete the escaped string
	// //
	//	delete [] safe_name;
	// 
	// // Perform the INSERT
	// //
	//	int newid = database->insertQuery(sqlstatement.c_str());
	//</code>
	// 
	//=
	static char *escape(const char *string, int length);
	
	//=
	// Returns the singleton database object for a given context
	//
	// <b>Example:</b>
	//
	// <code>
	// rude::Database *database = rude::Database::instance("context1");
	//
	// // use the database object...
	// </code>
	//=
	static Database *instance(const char *contextname);

	//=
	// Closes all open database connections
	// You can safely call this even if no connections were open and no instances were accessed.
	// We recommend that this method be called at the end of your main() method in any application 
	// that may use the database object.
	//=
	static void finish();

	//=
	// Returns the last known error
	//
	// <b>Example</b>
	// 
	// <code>
	// std::cout << database->getError() << "\n";
	// </code>
	//=
	const char *getError() const;
	
	//=
	// Returns MYSQL server status information
	//
	// <b>Example</b>
	//
	// <code>
	// std::cout << database->stat() << "\n";
	// </code>
	//
	//=
	const char *stat();
	
	//=
	// Sends a query that expects a result set (time friendly)
	// Obtains results from server all at once 
	// This is the faster alternative to calling useResultQuery(),
	// and is the preferred method when small result sets are expected. 
	// Returns false if an error occurs.
	// If it returns true, use nextRow() and column() to access the results.
	// If the first call to nextRow() returns false, then the result set is empty.
	// After you are done with the results, call freeResult() to free the system resources.
	//
	// <b>Example</b>
	//
	// <code>
	// if(database->storeResultQuery("SELECT username, password from tbl_users"))
	// {
	//    while(database->nextRow())
	//		{
	//			std::string username = database->column(0);
	//			std::string password = database->column(1);
	//       std::cout << "Username: << username << " Password: " << password << "\n";
	//		}
	//		database->freeResult();
	//	}
	//	else
	//	{
	//		std::cout << database->getError() << "\n";
	//	}
	// </code>
	//
	//=
	RUDEBOOL storeResultQuery(const char *query);

	//=
	// Sends a query that expects a result set (memory / network friendly)
	// Obtains results from server one row at a time 
	// This uses less memory and network resources than calling useResultQuery(),
	// and is the preferred method when large result sets are expected
	// Returns false if an error occurs.
	// If it returns true, use nextRow() and column() to access the results
	// If the first call to nextRow() returns false, then the result set is empty.
	// After you are done with the results, call freeResult() to free the system resources.
	// 
	// <b>Example</b>
	//
	// <code>
	// if(database->storeResultQuery("SELECT username, password from tbl_users"))
	// {
	//    while(database->nextRow())
	//		{
	//			std::string username = database->column(0);
	//			std::string password = database->column(1);
	//       std::cout << "Username: << username << " Password: " << password << "\n";
	//		}
	//		database->freeResult();
	//	}
	//	else
	//	{
	//		std::cout << database->getError() << "\n";
	//	}
	// </code>
	//
	//=
	RUDEBOOL useResultQuery(const char *query);

	//=
	// Selects the next row from query result set
	// Returns true if the next row is selected
	// Returns false if there are no more rows.
	// Must be called to get the first row.
	// 
	// <b>Example</b>
	//
	// <code>
	// if(database->storeResultQuery("SELECT category_name tbl_categories"))
	// {
	//    if(database->nextRow())
	//		{
	//			// We got some data back
	//			
	//			std::cout << "Current Categories:\n";
	//			do
	//			{
	//				std::cout << database->column(0) << "\n";
	//			}
	//			while(database->netxRow())
	//
	//		}
	//    else
	//		{
	//			// The result set was empty
	//
	//			std::cout << "There are no categories at this time\n";
	//		}
	//		database->freeResult();
	//	}
	//	else
	//	{
	//		std::cout << database->getError() << "\n";
	//	}
	// </code>
	//=
	RUDEBOOL nextRow();

	//=
	// Returns the value located at the specified column within the current row set
	// first index is 0.
	// will return NULL when out of range.
	// 
	// <b>Example</b>
	//
	// <code>
	// if(database->storeResultQuery("SELECT fname, lname, address, age from tbl_contacts"))
	// {
	//    while(database->nextRow())
	//		{
	//			const char *first_name = database->column(0);
	//			const char *last_name = database->column(1);
	//			const char *address = database->column(2);
	//			
	//			// convert the age to integer
	//			//
	//			int age = atoi( database->column(3) );
	//       
	//			// do stuff with data.....
	//
	//		}
	//		database->freeResult();
	//	}
	//	else
	//	{
	//		std::cout << database->getError() << "\n";
	//	}
	// </code>
	//
	//=
	const char *column(int index);

	//=
	// Frees resources consumed by a query result set
	// For every call to useResultQuery() or storeResultQuery() in your application, 
	// there should be an accompanying freeResult() call as well.
	//=
	RUDEBOOL freeResult();

	//=
	// Send a query that expects a single value result
	// Use this method when you expect a single value to be the result.
	// Returns c-string (const char *) representation of data if successful
	// Returns null if a result could not be obtained or an error occurs
	// <b>Example:</b>
	//
	// <code>
	// const char *username = db_database->singleValueQuery( "SELECT username form tbl_users where userid=348" );
	// if(username)
	// {
	//	   std::cout << "Username for givevn ID is " << username << "\n";
	// }
	// else
	// {
	//	   std::cout << "Userid 348 does not exist: " << db_database->getError() << "\n";
	// }
	// </code>	
	//=
	const char *singleValueQuery(const char *query);

	//=
	// Sends queries that do not return a result (UPDATE, DELETE, CREATE, INSERT, ALTER, etc...)
	// Returns true if successful, false otherwise.
	// NOTE: Use insertQuery() for INSERT's when you want to know the resulting insert-id.
	// 
	// <b>Example:</b>
	//
	// <code>
	// if( db_database->executeQuery( "DELETE FROM tbl_color WHERE color='blue'" ) )
	// {
	//	   std::cout << "Deleted the color blue"\n";
	// }
	// else
	// {
	//	   std::cout << db_database->getError() << "\n";
	// }
	// </code>	
	//=

RUDEBOOL executeQuery(const char *query);

	//=
	// Sends an INSERT query and returns the insert id
	// This is only applicable for tables with an AUTO_INCREMENT primary key.
	// For tables without an AUTO_INCREMENT primary key, use executeQuery() instead.
	// Returns 0 if an error occurs.
	// 
	// <b>Example:</b>
	//
	// <code>
	// if( int insertid = db_database->insertQuery( "INSERT INTO tbl_color SET color='blue'" ) )
	// {
	//	   std::cout << "Inserted a color with id of " << y << "\n";
	// }
	// else
	// {
	//	   std::cout << db_database->getError() << "\n";
	// }
	// </code>
	//=
	int insertQuery(const char *query);

	//=
	// 
	//=
	virtual ~Database();
};
}
#endif







